File: rust_ver/a/rust_fft_a/src/Complex.rs
pub struct Complex {
    real: f64,
    imag: f64
}

impl Complex {
    pub fn new(real: f64, imag: f64) -> Self {
        Self { real, imag }
    }

    pub fn add(&self, other: &Complex) -> Complex {
        Complex {
            real: self.real + other.real,
            imag: self.imag + other.imag,
        }
    }

    pub fn sub(&self, other: &Complex) -> Complex {
        Complex {
            real: self.real - other.real,
            imag: self.imag - other.imag,
        }
    }

    pub fn mul(&self, other: &Complex) -> Complex {
        Complex {
            real: self.real * other.real - self.imag * other.imag,
            imag: self.real * other.imag + self.imag * other.real,
        }
    }

    pub fn div(&self, other: &Complex) -> Complex {
        denom: f64 = other.real * other.real + other.imag * other.imag;
        numer_real: f64 = self.real * other.real + self.imag * other.imag;
        numer_imag: f64 = self.imag * other.real - self.real * other.imag
        Complex {
            real: numer_real / denom,
            imag: numer_imag / denom
        }
    }

    pub fn integer_pow(&self, int times) {
        if times == 0 {
            return Complex::new(1.0, 0.0);
        } else {
            let mut result = Complex::new(self.real, self.imag);
            for _ in 0..times {
                result = result.mul(self);
            }
            return result;
        }
    }
}

File: java_ver/fft_a/Complex.java
import java.lang.Math;

public class Complex {
    double real;
    double imag;

    Complex(double real, double imag) {
        this.real = real;
        this.imag = imag;
    }

    /**
     * @param a is module or real
     * @param b is radian or imag
     * @param ccs tells java how to resolution the parameters
     */
    public Complex(double a, double b, boolean ccs) {
        if (ccs) {
            this.real = a;
            this.imag = b;
        } else {
            this.real = a * Math.cos(b);
            this.imag = a * Math.sin(b);
        }
    }

    public void SetInCCS(double real, double imag) {
        this.real = real;
        this.imag = imag;
    }

    public void SetInPCS(double mod, double angle) {
        this.real = mod * Math.cos(angle);
        this.imag = mod * Math.sin(angle);
    }

    public void Add(Complex b) {
        this.real += b.real;
        this.imag += b.imag;
    }

    public void Min(Complex b) {
        this.real -= b.real;
        this.imag -= b.imag;
    }

    //  (a + bi)(c + di)
    // =(ac - bd) + (ad + bc)i
    public void Mul(Complex b) {
        double real = (this.real * b.real) - (this.imag * b.imag);
        double imag = (this.real * b.imag) + (this.imag * b.real);
        this.real = real;
        this.imag = imag;
    }

    //  (a + bi) / (c + di)
    // =(a + bi)(c - di) / (c + di)(c - di)
    // =((ac + bd) + (bc - ad)i) / (c^2 + d^2)
    // =((ac + bd) / (c^2 + d^2)) + ((bc - ad) / (c^2 + d^2))i
    public void Div(Complex b) {
        double denom = (b.real * b.real) + (b.imag * b.imag);
        double numer_real = (this.real * b.real) + (this.imag * b.imag);
        double numer_imag = (this.imag * b.real) - (this.real * b.imag);
        this.real = numer_real / denom;
        this.imag = numer_imag / denom;
    }

    public void IntegerPow(int times) {
        if (times == 0) {
            this.real = 1;
            this.imag = 0;
        } else {
            Complex q = new Complex(this.real, this.imag);
            for (int i = 0; i < times; i++) {
                this.Mul(q);
            }
        }
    }

    @Override
    public String toString() {
        return "Complex[real = " + this.real + ", imag = " + this.imag + "]";       
    }

    public double getReal() {
        return this.real;
    }

    public double getImag() {
        return this.imag;
    }
}